pkg/component/gardenerapiserver/gardener_apiserver.go

```diff
@@ -0,0 +1,205 @@
+func New(client client.Client, namespace string, secretsManager secretsmanager.Interface, values Values) Interface {
+	return &gardenerAPIServer{
+		client:         client,
+		namespace:      namespace,
+		secretsManager: secretsManager,
+		values:         values,
+	}
+}
+
+type gardenerAPIServer struct {
+	client         client.Client
+	namespace      string
+	secretsManager secretsmanager.Interface
+	values         Values
+}
+
+func (g *gardenerAPIServer) Deploy(ctx context.Context) error {
+	var (
+		runtimeRegistry = managedresources.NewRegistry(operatorclient.RuntimeScheme, operatorclient.RuntimeCodec, operatorclient.RuntimeSerializer)
+	)
+
+	runtimeResources, err := runtimeRegistry.AddAllAndSerialize()
+	if err != nil {
+		return err
+	}
+
+	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	if err := managedresources.CreateForSeed(ctx, g.client, g.namespace, managedResourceNameRuntime, false, runtimeResources); err != nil {
+		return err
+	}
+	if err := managedresources.WaitUntilHealthy(timeoutCtx, g.client, g.namespace, managedResourceNameRuntime); err != nil {
+		return err
+	}
+
+	var (
+		virtualRegistry = managedresources.NewRegistry(operatorclient.VirtualScheme, operatorclient.VirtualCodec, operatorclient.VirtualSerializer)
+	)
+
+	virtualResources, err := virtualRegistry.AddAllAndSerialize()
+	if err != nil {
+		return err
+	}
+
+	timeoutCtx, cancel = context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	if err := managedresources.CreateForShoot(ctx, g.client, g.namespace, managedResourceNameVirtual, managedresources.LabelValueGardener, false, virtualResources); err != nil {
+		return err
+	}
+	return managedresources.WaitUntilHealthy(timeoutCtx, g.client, g.namespace, managedResourceNameVirtual)
+}
+
+func (g *gardenerAPIServer) Destroy(ctx context.Context) error {
+	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	if err := managedresources.DeleteForShoot(ctx, g.client, g.namespace, managedResourceNameVirtual); err != nil {
+		return err
+	}
+	if err := managedresources.WaitUntilDeleted(timeoutCtx, g.client, g.namespace, managedResourceNameVirtual); err != nil {
+		return err
+	}
+
+	timeoutCtx, cancel = context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	if err := managedresources.DeleteForSeed(ctx, g.client, g.namespace, managedResourceNameRuntime); err != nil {
+		return err
+	}
+	return managedresources.WaitUntilDeleted(timeoutCtx, g.client, g.namespace, managedResourceNameRuntime)
+}
+
+func (g *gardenerAPIServer) Wait(ctx context.Context) error {
+	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	if err := g.waitUntilRuntimeManagedResourceHealthyAndNotProgressing(ctx); err != nil {
+		return err
+	}
+	return managedresources.WaitUntilHealthy(timeoutCtx, g.client, g.namespace, managedResourceNameVirtual)
```

Comment:
What is the reason for having different checks for runtime and virtual managed resource here? In other words, why is `managedresources.WaitUntilHealthy(...)` not used for the runtime managed resource?

Reply:
Typically, we use `managedresources.WaitUntilHealthy` by default everywhere/in all components.
However, here we have to wait for the runtime resources to no longer be progressing before we can update the virtual resources. This is important for CA rotation since we want all GAPI pods to run with the new server certificate before we drop the old CA from the bundle in the `APIService`s (which get deployed via the virtual resources). See also https://sap-ti.slack.com/archives/C04SLGN4L2V/p1687506600947999?thread_ts=1686837742.192359&cid=C04SLGN4L2V.

Reply:
Thanks for the explanation.